components/service/glean/src/main/java/mozilla/components/service/glean/storages/PingStorageEngine.kt

```diff
@@ -156,6 +149,7 @@ internal class PingStorageEngine(context: Context) {
      * @param pingPath The upload path to serialize
      * @param pingData The ping data to serialize
      */
+    @Synchronized
```

Comment:
What happens if we trigger on the number of events at the same time the app is going to background, we could have two processes trying to write to the same file, right?  I suppose I could move this to just wrap around the event ping serialization stuff.

Reply:
Wouldn't they have a different filename, given that different pings have different UUIDs? I could have also just misunderstood the problem :)

Reply:
Think of this situation...
1) We have an event tied to the 'back' button (which could potentially put the app in the background)
2) The clicking of the 'back' button triggers the 500 event threshold to send the event ping, as well as going into the background which also triggers the event ping

Couldn't they possibly interfere with each other?  

I've removed this and kept it the same for now... but it was a concern I had while thinking through it...

Reply:
> 
> 
> Think of this situation...

Nice! I like talking through examples, thanks for expanding on this!

>     1. We have an event tied to the 'back' button (which could potentially put the app in the background)
> 
>     2. The clicking of the 'back' button triggers the 500 event threshold to send the event ping, as well as going into the background which also triggers the event ping
> 
> 
> Couldn't they possibly interfere with each other?

I think it shouldn't. If it does, it's a bug in some component up higher in our glean hierarchy. Recording an event is a `synchronized` operation, because of the storage engine implementation. Recording the 500th event would trigger a collection, generate the ping, empty the storage. This event would be saved to `uuid1.json`. The next record would accumulate in an empty storage. Going to background would again trigger collection, generate the ping and empty the storage egain. This would be saved to `uuid2.json`. Is this the scenario you were thinking about?

Reply:
Yes, but could we get two pings (with different uuid filenames) that contain the same events?

Reply:
Sorry, thought experiment that I can probably execute myself.  Just thinking out loud :)

Reply:
> 
> 
> Yes, but could we get two pings (with different uuid filenames) that contain the same events?

They shouldn't contain the same events. If they do, then it's a bug in the `EventsStorageEngine`. Did you see this in some test/manual run?

Reply:
> 
> 
> Sorry, thought experiment that I can probably execute myself. Just thinking out loud :)

No problem, I'm *very* happy to talk through examples!

Reply:
Nope, I haven't.  Ultimately, you are right I added this to fight the ghost of a bug I've never seen :)